### PR TITLE
WIP | Dev: Initial bit integration

### LINKED ISSUE #1028: Dev: Initial BIT-MS integration - CORS and Postgresql | CROSS PROJECT ISSUE
## Description
As a contributor and the BridgeInTech maintainer,
I need to integrate the mentorship backend version for BridgeInTech project to the main Mentorship System repository,
so that I can invite more contributors to work on the BIT-MS integration related issues.

## Mocks
NA

## Acceptance Criteria
- [ ] Add CORS to allow communication between MS backend and BIT backend servers
- [ ] Change DB to use only postgresql since it allows multiple schemas in one database.

### Update [Required]
- [ ] update dependencies by adding Flask-Cors==3.0.9

### Enhancement to Update [Optional]
- [ ] NA

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
1 hour

### PR DESCRIPTION
**Pull request by mtreacy002**
### Description
Add initial integration for BIT backend to MS backend 

Fixes #1028 

### Type of Change:
- Code
- Quality Assurance

**Code/Quality Assurance Only**
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
1. Complete the local development setup as instructed on [BIT wiki page](https://github.com/anitab-org/bridge-in-tech-backend/blob/develop/.github/ENV_SETUP_INSTRUCTION.md). For ms-backend-server bit version, run it from this PR branch.
2. run BIT backend server FIRST to create the schemas inside postgresql. then run MS backend server from this PR branch.
3. Register a user and see success message.
Screen Shot 2021-03-03 at 5 39 20 pm
4. run tests by executing the command `python -m unittest discover tests` and see all test run successfully
Screen Shot 2021-03-03 at 5 49 24 pm

5. Login on BIT Swagger UI and make call requests related to MS users service (on gif below GET user/personal_details and GET users/{user_id} were made as examples)
ezgif com-gif-maker (36)

### Checklist:
- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] Any dependent changes have been merged
- [ ] Update requirements.txt

**Code/Quality Assurance Only**

- [ ] My changes generate no new warnings 
- [ ] New and existing unit tests pass locally with my changes

**Additional note**
For the giithub action to run test successfully, the `main.yml` need to be modified to include adding postgresql plugin and creating test database with the 2 schemas as it was done on Travis.yml on BIT backend repository.
Screen Shot 2021-03-03 at 5 55 38 pm

### CODE REVIEW

File: app/database/models/mentorship_relation.py
Comment (epicadk) [inline app/database/models/mentorship_relation.py:21]: I honestly don't understand why we use floating point numbers for dates on the mentorship system backend to being with. Should we just add this to the develop branch as well?

File: config.py
Comment (epicadk) [inline config.py:44]: We have fetched the environment variable here .... (continued below)

File: config.py
Comment (epicadk) [inline config.py]: Then why not use the DB_NAME variable instead of refetching them from the environment variables?

File: run.py
Comment (epicadk) [inline run.py:29]: Does BIT required access to all endpoints of the mentorship backend or can we give access to specific endpoints instead of all the endpoints?

File: app/database/models/mentorship_relation.py
Comment (mtreacy002) [reply app/database/models/mentorship_relation.py:21]: I understand your point, but this is for @isabelcosta to decide 😉.

File: config.py
Comment (mtreacy002) [reply config.py:44]: @epicadk , the one below is for different db_uri (test db_uri). This one is for the main db_uri). The thing is, I'm aiming for better code readability then trying to put them together but makes it less readable 😁.

File: config.py
Comment (mtreacy002) [reply config.py:86]: I'm not sure I'm following, 🤔.

Comment (epicadk) [reply config.py]: > I'm not sure I'm following, 🤔.

This was a continuation from the above comment. On line 44 there is a variable called DB_NAME that has this value, then why not use that here instead of fetching the environment variable again.

File: run.py
Comment (mtreacy002) [reply run.py:29]: Yes, @epicadk. With users functionaliities and mentorship relation are managed by Mentorship System, BIT does need to have access to all Mentoship System api endpoints 😉.

File: config.py
Comment (mtreacy002) [reply config.py:44]: I'll try to refine the code now 😉

File: config.py
Comment (mtreacy002) [reply config.py]: Fixing it now. Thanks for pointing that out 😊

Comment (mtreacy002) [reply config.py]: @epicadk , done. Can you please check again? Thanks

File: run.py
Comment (epicadk) [reply run.py:29]: Do you think we should fetch the origins property based on the build type i.e. production, dev, local etc.?

File: config.py
Comment (epicadk) [reply config.py]: Looks good 💯

**Comment by mtreacy002** [comment]
@epicadk and @annabauza . I'm trying to set up postgresql inside github action but having trouble getting it right. 
Screen Shot 2021-03-04 at 12 42 22 am

Here's the modification to [`main.yml`](https://github.com/mtreacy002/mentorship-backend/blob/initial-bit-integration/.github/workflows/main.yml) i've done.
From the PR description above (`Additional Content`) you can see how I've done it using Travis CI
PS: I've never written CI/CD to github action before so, any help would be appreciated.

**Comment by epicadk** [comment]
It won't let you create a database for some reason it is setup just fine. Is there anyone to set these from the python code?

**Comment by mtreacy002** [comment]
> It won't let you create a database for some reason it is setup just fine. Is there anyone to set these from the python code?

can you explain what you mean by setting up on the python code, @epicadk ?

**Comment by epicadk** [comment]
> > It won't let you create a database for some reason it is setup just fine. Is there anyone to set these from the python code?
> 
> can you explain what you mean by setting up on the python code, @epicadk ?

Like we do for the mentorship system `db.create_all()` over [here](https://github.com/anitab-org/mentorship-backend/blob/479f1c75a8a84a598612ef53f7bb7bc372e13f3e/run.py#L45)

**Comment by mtreacy002** [comment]
> > > It won't let you create a database for some reason it is setup just fine. Is there anyone to set these from the python code?
> > 
> > 
> > can you explain what you mean by setting up on the python code, @epicadk ?
> 
> Like we do for the mentorship system `db.create_all()` over [here](https://github.com/anitab-org/mentorship-backend/blob/479f1c75a8a84a598612ef53f7bb7bc372e13f3e/run.py#L45)

I've done that, @epicadk , but still facing error 😓

**Comment by mtreacy002** [comment]
Update @epicadk :
I'm now able to create test db and connect to Postgres in the build 🎉 💃 
But now I'm stuck with `install dependencies` as shown below. Any idea why this happens? Does this have something to do with permission level to the Mentorship System GitHub repo? cc @isabelcosta 
Screen Shot 2021-03-05 at 12 02 34 am

I also noticed that when the PR #654 for initial MS migration to GitHub action was merged, the `Run test` workflow did not get triggered. It was only triggered on the PR after  (#655). I wonder why 🤔...
Screen Shot 2021-03-05 at 12 24 02 am

**Comment by annabauza** [comment]
Maya github have environment variable you can setup, but I don't think you need it here, you should define them on heroku itself

**Comment by mtreacy002** [comment]
> Maya github have environment variable you can setup, but I don't think you need it here, you should define them on heroku itself

@annabauza , are you talking about this error message?  
Screen Shot 2021-03-05 at 12 02 34 am

**Comment by annabauza** [comment]
Ahh sorry I forgot you have testing database... Ignore me please.

File: run.py
Comment (mtreacy002) [reply run.py:29]: @epicadk, I think it's best to keep it as it for now, coz majority of contributors work on local machine atm. When we have remote servers for dev and production, we can revisit this origin property.

**Comment by NenadPantelic** [comment]
@mtreacy002 Try with this:
`pip3 install --no-cache-dir -r requirements.txt` instead of the current pip installation command. I think it is a user permission error.

**Comment by epicadk** [comment]
> @mtreacy002 Try with this:
> `pip3 install --no-cache-dir -r requirements.txt` instead of the current pip installation command. I think it is a user permission error.

Yeah that error was fixed. If you check @mtreacy002 's fork I have sent a pr however now the tests are failing. I think the error may be in the database uri that we are providing.

**Comment by mtreacy002** [comment]
Hi @epicadk , I want to show that you've co-authored this pr. How to do this? Should I change manually by replacing name and email address below? If so, can you give your email address here?
Screen Shot 2021-03-06 at 10 18 38 am

**Comment by mtreacy002** [comment]
@epicadk , the tests and lint are now passing 🎉 💃 
But the codecov still fails 😓
How do we fix this? Can we remove the codecov from `main.yml` since the repo will do this separately as this is how it is done on `develop` branch.

**Comment by mtreacy002** [comment]
All tests passed now 🎉 💃 
@epicadk can you please review again? Thanks

**Comment by epicadk** [comment]
> Hi @epicadk , I want to show that you've co-authored this pr. How to do this? Should I change manually by replacing name and email address below? If so, can you give your email address here?
> Screen Shot 2021-03-06 at 10 18 38 am

Haha that's okay once it'll probably get lost in the squashed commits. Glad I could help. 😁

File: config.py
Comment (epicadk) [inline config.py]: We are returning a hard coded string. Maybe if we return a string based using the environment variables, contributors won't have to uncomment this. Do you think that is a good idea?

Comment (mtreacy002) [reply config.py]: Yep, fix it now by using the existing db_test_uri for non-MacOS lines 😉

**Review by epicadk — APPROVED** [review]
LGTM 👍🏼 . But I have a question about the testing. Will I need to setup BIT on my local machine as well or can I just run the tests on the mentorship system backend?

**Comment by mtreacy002** [comment]
> LGTM 👍🏼 . But I have a question about the testing. Will I need to setup BIT on my local machine as well or can I just run the tests on the mentorship system backend?

If this PR is merged, we can use this branch as the MS backend server for BIT project, instead of setting up a different MS backend from my fork repo. But as BIT project has 3 servers, then yes, you would still need to setup BIT backend so you can test the BIT features. Hope this answer your question 😸 
PS: The condition of BIT backend server must be run first before running the MS backend server on initial setup of BIT project still applies, so the Postgres db gets created properly with the 2 schemas from BIT backend code.

File: app/api/resources/user.py
Comment (isabelcosta) [inline app/api/resources/user.py]: 😝 can you remove this print() statement?

Comment (isabelcosta) [reply app/api/resources/user.py]:
```suggestion
```
Or is it needed @mtreacy002 ?

Comment (isabelcosta) [reply app/api/resources/user.py]: if it's not needed, just remove it and I am happy to merge it after @mtreacy002

Comment (mtreacy002) [reply app/api/resources/user.py]: Ooppss, forgot to take it out 😂. Will remove it asap, @isabelcosta 😉

Comment (mtreacy002) [reply app/api/resources/user.py]: Done, @isabelcosta , please review again when you have time 😉

File: .github/workflows/main.yml
Comment (devkapilbansal) [inline .github/workflows/main.yml]: This line is not useful

File: app/api/dao/user.py
Comment (devkapilbansal) [inline app/api/dao/user.py:262]: Why this field is renamed? Is it for making it compatible with `BIT`

File: app/api/models/user.py
Comment (devkapilbansal) [inline app/api/models/user.py]:
```suggestion
            required=True, description="User current organization"
```

Comment (devkapilbansal) [reply app/api/models/user.py]: It would be better if same convention is followed in `models/admin.py` and `models/user.py`.

Comment (devkapilbansal) [reply app/api/models/user.py]: Same for other instances as well

File: config.py
Comment (devkapilbansal) [inline config.py:86]: If we are not using these variables, then why they are used in function definition?

File: tests/test_app_config.py
Comment (devkapilbansal) [inline tests/test_app_config.py:27]: Can we add similar line for `postgres`?

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
@mtreacy002 although the code looks good, I have some questions/suggestions as follows:-

File: .github/workflows/main.yml
Comment (devkapilbansal) [reply .github/workflows/main.yml]: For line no. 56 only

Comment (epicadk) [reply .github/workflows/main.yml]: > For line no. 56 only

Oops sorry my bad I added that when it wasn't working just to make sure pip is working.

File: tests/test_app_config.py
Comment (epicadk) [reply tests/test_app_config.py:27]: I think it would depend on whether the developer is using `postgres` or `postgres+psycopg2`.

File: app/api/dao/user.py
Comment (mtreacy002) [reply app/api/dao/user.py:262]: that's right, @devkapilbansal . because bit schema has an attribute called organization as well. So, I renamed it here on MS coz it clashes with the BIT attribute otherwise.

File: tests/test_app_config.py
Comment (mtreacy002) [reply tests/test_app_config.py:27]: @devkapilbansal , as per @epicadk said, this is very machine specific. To place it here in the test config means we have to provide separate syntax for each machines. Plus, we will ask contributors to always comment in/out the line/s that are not relevant to their machine when they run the local testing. After that, when they push the PR, they have to always only comment in the line that are going to be used when running test on CI (which is using ubuntu-latest) and comment out the rest, so, I don't think it's ideal to have the db_uri matching syntax to be added to the test.
Unfortunately, atm this is how we do it in the `config.py` for Local and Test config for different OS. Perhaps you can suggest a logic to tackle this issue?

File: config.py
Comment (mtreacy002) [reply config.py:86]: this variables is being used for the linux/windows os, by uncommenting line 131 below and comment out line 129

File: app/api/models/user.py
Comment (mtreacy002) [reply app/api/models/user.py:51]: > It would be better if same convention is followed in `models/admin.py` and `models/user.py`.

models/admin, @devkapilbansal ? I can't see that file exist in MS backend base code?

> Same for other instances as well

I'm not sure I'm following, can you please elaborate? Which convention you are talking here?

File: config.py
Comment (mtreacy002) [reply config.py:86]: this variables is being used for the linux/windows os, by uncommenting line 131 below and comment out line 129

File: .github/workflows/main.yml
Comment (mtreacy002) [reply .github/workflows/main.yml]: Done. I've removed it in the latest push 😉

**Comment by mtreacy002** [comment]
@devkapilbansal . Can you please review again?

File: app/api/models/user.py
Comment (mtreacy002) [reply app/api/models/user.py]: done

**Comment by mtreacy002** [comment]
@epicadk, can you please tell me how to fix the failed codecov? Also, why is the codecov report is not showing for this PR just like it is shown on the other PR for `develop` branch (e.g. #1039)?

**Comment by epicadk** [comment]
@mtreacy002 could you avoid force pushing the changes? Kinda makes it harder to review. I had discussed this with @isabelcosta in the QA open session and I think the conclusion was that only while merging the maintainer needs to follow the commit guidelines given in the contributing guidelines. cc @isabelcosta . 😅

**Comment by mtreacy002** [comment]
> @mtreacy002 could you avoid force pushing the changes? Kinda makes it harder to review. I had discussed this with @isabelcosta in the QA open session and I think the conclusion was that only while merging the maintainer needs to follow the commit guidelines given in the contributing guidelines. cc @isabelcosta . 😅

Ok, it makes sense actually 😁. It's just a habit from squashing commits, I guess (which I actually forgot to do on the latest push 🤣). Will just do normal push the next one 😉

**Comment by mtreacy002** [comment]
Oh, on the codecov report, I check the [anitab codecov.io](https://app.codecov.io/gh/anitab-org/mentorship-backend/pulls?page=1&state=open&order=-pullid), and it looks like it's passing, so I'm not sure why it says failed here 🤔
Screen Shot 2021-03-14 at 12 38 22 am

**Comment by epicadk** [comment]
I don't have much knowledge of the subject 😅. I think @devkapilbansal can help out here.

**Comment by mtreacy002** [comment]
> I don't have much knowledge of the subject 😅. I think @devkapilbansal can help out here.

no worries..., @epicadk 😉. @devkapilbansal can you please share some input 🙏.

**Review by devkapilbansal — COMMENTED** [review]
@mtreacy002 this PR consists of 5 commits but only one is relevant. All four are possibly because `bit` branch is not updated with `develop` branch. 
So they should be removed.
Otherwise it is good for me.

**Comment by mtreacy002** [comment]
> @mtreacy002 this PR consists of 5 commits but only one is relevant. All four are possibly because `bit` branch is not updated with `develop` branch.
> So they should be removed.
> Otherwise it is good for me.

Done, @devkapilbansal , I removed the other 4 commits. It was initially to update `bit` branch with develop. But the maintainers can do that from their side too 😉

**Comment by mtreacy002** [comment]
@epicadk , should I also integrate your push for docker change in here? Or that can be done after?

**Comment by epicadk** [comment]
> @epicadk , should I also integrate your push for docker change in here? Or that can be done after?

It can be done after. I'll send open the pr on this repository.

**Comment by mtreacy002** [comment]
@devkapilbansal and @epicadk , can you please re-approve this PR now so it can be merged? That way the contributors can push directly to upstream `bit` branch instead of to my fork ms repo. Thanks again 😉

**Review by epicadk — APPROVED** [review]
💯

**Review by isabelcosta — APPROVED** [review]
Looks good to me! @mtreacy002 🤗

**Comment by isabelcosta** [comment]
@mtreacy002 this is working for your specific case right? Should we test this again? or pass that, because you were already using this for BIT?

**Review by devkapilbansal — APPROVED** [review]
LGTM :+1: 

P.S.:- I haven't tested it

**Comment by devkapilbansal** [comment]
> @epicadk, can you please tell me how to fix the failed codecov? Also, why is the codecov report is not showing for this PR just like it is shown on the other PR for `develop` branch (e.g. #1039)?

@mtreacy002 @epicadk looking into the codecov reports, it fails because of changes in `run.py` file. Also, in workflows, branches mentioned is `develop`. That's why codecov report is not shown here

**Comment by mtreacy002** [comment]
> @mtreacy002 this is working for your specific case right? Should we test this again? or pass that, because you were already using this for BIT?

@isabelcosta , I believe since the GitHub action shows that the build is passing here we won't need to do the manual testing again. I also have been using this branch for my BIT for MS backend server and able to use existing features as normal.
PS: I've added a gif on successful calls to MS users REST API from BIT Swagger UI using this PR branch under `How has this been tested` section.


> > @epicadk, can you please tell me how to fix the failed codecov? Also, why is the codecov report is not showing for this PR just like it is shown on the other PR for `develop` branch (e.g. #1039)?
> 
> @mtreacy002 @epicadk looking into the codecov reports, it fails because of changes in `run.py` file. Also, in workflows, branches mentioned is `develop`. That's why codecov report is not shown here

@devkapilbansal, can you tell me where the `develop` branches being mentioned in workflows that causing the codecov fails? as I've changed this to `bit` in the `main.yml` on this pr.
Screen Shot 2021-03-15 at 12 17 18 pm

Are you talking on the codecov.io settings of AnitaB? Perhaps this must be done from the admins side since they have the authority on that?

**Comment by mtreacy002** [comment]
@devkapilbansal , @epicadk , can you please let me know what else I need to do for this PR? If nothing else needed, we can label this issue `Ready to merge` so the maintainer can merge it to the MS `bit` branch.

**Comment by epicadk** [comment]
> @devkapilbansal , @epicadk , can you please let me know what else I need to do for this PR? If nothing else needed, we can label this issue `Ready to merge` so the maintainer can merge it to the MS `bit` branch.

We have code approval . Now it needs testing . Busy with college this week i'll test it tonight.

**Comment by epicadk** [comment]
Okay I just realized I'll need to have postgreSQL installed to test this which I haven't installed yet.

**Comment by mtreacy002** [comment]
Will close this issue as we're moving to full BIT-MS integration as discussed on BIT Open Session Sunday, 11th April.

**Pull request closed without merging**
